KubeArmor/monitor/systemMonitor.go

```diff
@@ -152,9 +158,138 @@ func NewSystemMonitor(node *tp.Node, logger *fd.Feeder, containers *map[string]t
+	log.PPID = int32(pidInfo.PPID)
+	log.UID = int32(pidInfo.UID)
+
+	log.Source = pidInfo.Source
+	if len(pidInfo.Args) > 0 {
+		log.Source = log.Source + " " + pidInfo.Args
+	}
+
+	log.ParentProcessName = pidInfo.ParentExecPath
+	log.ProcessName = pidInfo.ExecPath
+
+	if match[1] == "DENIED" {
+		log.Result = "Permission denied"
+	} else {
+		log.Result = "Passed"
+	}
+
+	log.Data = "operation=" + match[2]
+
+	switch match[2] {
+	case "exec":
+		log.Operation = "Process"
+		log.Resource = match[3]
+	case "open", "getattr":
+		log.Operation = "File"
+		log.Resource = match[3]
+	case "create", "connect", "setsockopt", "getpeername", "sendmsg", "recvmsg":
+		log.Operation = "Network"
+		proto, _ := strconv.ParseInt(match[6], 10, 32)
+		log.Resource = "domain=" + match[4] + " type=" + match[5] + " protocol=" + getProtocol(int32(proto))
+	case "capable":
+		log.Operation = "Network"
+		log.Resource = "capability_id=" + match[4] + " capability_name=" + match[5]
+	default:
+		mon.Logger.Warnf("Dropped AppArmor Alert, no matching operation: %v", match)
+		return
+	}
+
+	go mon.Logger.PushLog(log)
+}
+
+func (mon *SystemMonitor) WatchAppArmorAlerts() {
+	if mon.NetLinkClient == nil {
+		return
+	}
+	ActiveHostPidMap := *(mon.ActiveHostPidMap)
+	ActivePidMapLock := *(mon.ActivePidMapLock)
+
+	profilergx := regexp.MustCompile("apparmor=\"(AUDIT|DENIED)\".*?operation=\"(.*?)\".*?name=\"(.*?)\".*?pid=([0-9]+)")
+	netregx := regexp.MustCompile("apparmor=\"(AUDIT|DENIED)\".*?operation=\"(.*?)\".*?pid=([0-9]+).*?family=\"(.*?)\".*?sock_type=\"(.*?)\" protocol=(.*?) ")
+	capregx := regexp.MustCompile("apparmor=\"(AUDIT|DENIED)\".*?operation=\"(.*?)\".*?pid=([0-9]+).*?capability=([0-9]+).*?capname=\"(.*?)\"")
+	for {
+		rawEvent, err := mon.NetLinkClient.Receive(false)
+		if err != nil || rawEvent.Type != auparse.AUDIT_AVC {
+			continue
+		}
+
+		if match := profilergx.FindAllStringSubmatch(string(rawEvent.Data), -1); len(match) != 0 {
+			ActivePidMapLock.Lock()
+			a, _ := strconv.Atoi(match[0][4])
+			for cID, pidInfor := range ActiveHostPidMap {
+				if elem, ok := pidInfor[uint32(a)]; ok {
+					go mon.BuildAppArmorLogBase(cID, elem, match[0])
+					break
+				}
+			}
+			ActivePidMapLock.Unlock()
+		} else if match := netregx.FindAllStringSubmatch(string(rawEvent.Data), -1); len(match) != 0 {
+
+			ActivePidMapLock.Lock()
+			a, _ := strconv.Atoi(match[0][3])
+			for cID, pidInfor := range ActiveHostPidMap {
+				if elem, ok := pidInfor[uint32(a)]; ok {
+					go mon.BuildAppArmorLogBase(cID, elem, match[0])
+					break
+				}
+			}
+			ActivePidMapLock.Unlock()
+		} else if match := capregx.FindAllStringSubmatch(string(rawEvent.Data), -1); len(match) != 0 {
+
+			ActivePidMapLock.Lock()
+			a, _ := strconv.Atoi(match[0][3])
+			for cID, pidInfor := range ActiveHostPidMap {
+				if elem, ok := pidInfor[uint32(a)]; ok {
+					go mon.BuildAppArmorLogBase(cID, elem, match[0])
+					break
+				}
+			}
+			ActivePidMapLock.Unlock()
+		}
```

Comment:
Can we put the lock/unlock outside the if-else-if chain?
```suggestion

		ActivePidMapLock.Lock()
		if match := profilergx.FindAllStringSubmatch(string(rawEvent.Data), -1); len(match) != 0 {
			a, _ := strconv.Atoi(match[0][4])
			for cID, pidInfor := range ActiveHostPidMap {
				if elem, ok := pidInfor[uint32(a)]; ok {
					go mon.BuildAppArmorLogBase(cID, elem, match[0])
					break
				}
			}
		} else if match := netregx.FindAllStringSubmatch(string(rawEvent.Data), -1); len(match) != 0 {
			a, _ := strconv.Atoi(match[0][3])
			for cID, pidInfor := range ActiveHostPidMap {
				if elem, ok := pidInfor[uint32(a)]; ok {
					go mon.BuildAppArmorLogBase(cID, elem, match[0])
					break
				}
			}
		} else if match := capregx.FindAllStringSubmatch(string(rawEvent.Data), -1); len(match) != 0 {
			a, _ := strconv.Atoi(match[0][3])
			for cID, pidInfor := range ActiveHostPidMap {
				if elem, ok := pidInfor[uint32(a)]; ok {
					go mon.BuildAppArmorLogBase(cID, elem, match[0])
					break
				}
			}
		}
		ActivePidMapLock.Unlock()
```